src/logger.ts

```diff
@@ -0,0 +1,432 @@
+import * as Bluebird from 'bluebird';
+import * as es from 'event-stream';
+import { ClientRequest } from 'http';
+import * as https from 'https';
+import * as _ from 'lodash';
+import * as Lock from 'rwlock';
+import * as stream from 'stream';
+import * as url from 'url';
+import * as zlib from 'zlib';
+
+import { EventTracker } from './event-tracker';
+import Docker = require('./lib/docker-utils');
+import { LogType } from './lib/log-types';
+
+const ZLIB_TIMEOUT = 100;
+const COOLDOWN_PERIOD = 5 * 1000;
+const KEEPALIVE_TIMEOUT = 60 * 1000;
+const RESPONSE_GRACE_PERIOD = 5 * 1000;
+
+const MAX_LOG_LENGTH = 10 * 1000;
+const MAX_PENDING_BYTES = 256 * 1024;
+
+interface Options extends url.UrlWithParsedQuery {
+	method: string;
+	headers: Dictionary<string>;
+}
+
+type LogMessage = Dictionary<any>;
+
+abstract class LogBackend {
+	public offlineMode: boolean;
+	public publishEnabled: boolean = true;
+
+	public abstract log(message: Dictionary<any>): void;
```

Comment:
I think this should be `message: LogMessage` - I know they're the same but changing it means that when `LogMessage` gets tightened then this automatically does as well

Reply:
Agreed, good catch